### PR TITLE
Refactor template system

### PR DESCRIPTION
**Pull request by gonzaloriestra**
### WHY are these changes introduced?

Fixes https://github.com/Shopify/internal-cli-foundations/issues/628

### WHAT is this pull request doing?

- Load all the extensions from template specifications: functions from the server, UI + themes locally (for now)

### How to test your changes?

`p shopify app generate extension`

### Measuring impact

How do we know this change was effective? Please choose one:

- [ ] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [x] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible [documentation](https://shopify.dev) changes
- [x] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

File: packages/app/src/cli/models/templates/ui-specifications/checkout_ui_extension.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        {
+          name: 'TypeScript React',
+          value: 'typescript-react',
+          path: 'packages/app/templates/ui-extensions/projects/checkout_ui_extension',
+        },
+        {
+          name: 'JavaScript React',
+          value: 'react',
+          path: 'packages/app/templates/ui-extensions/projects/checkout_ui_extension',
+        },
```
Comment (isaacroldan) [inline packages/app/src/cli/models/templates/ui-specifications/checkout_ui_extension.ts:36]: Since all UI templates share the same flavors, can we extract this to a shared function?
```ts
uiFlavors(path: string) {
...
}
```

File: packages/app/src/cli/services/generate.ts [hunk 2]
```diff
@@ -90,31 +79,36 @@ async function buildPromptOptions(
...
-  if (isFunctionExtension(type)) {
+function limitReached(app: AppInterface, specifications: GenericSpecification[], templateType: TemplateType) {
+  const type = templateType.type
+  if (type === 'function') {
     return app.extensions.function.length >= blocks.functions.defaultRegistrationLimit
   } else {
-    return app.extensionsForType(type).length >= type.registrationLimit
+    const specification = specifications.find((spec) => spec.identifier === type || spec.externalIdentifier === type)
+    const existingExtensions = app.extensionsForType({identifier: type, externalIdentifier: type})
+    return existingExtensions.length >= (specification?.registrationLimit || 1)
```
[the remark this answers is not in the document]
Comment (gonzaloriestra) [reply packages/app/src/cli/services/generate.ts:101]: You could be over the limit for different reasons (for example, if you generate two extensions at the same time in two different tabs). How we handle deploys or other stuff is another story, but here we should just avoid creating more extensions when you have more than the limit. So I think >= is safer than =.

File: packages/app/src/cli/services/generate.ts [hunk 3]
```diff
@@ -44,44 +39,38 @@ export interface GenerateOptions {
 }
```
[the remark this answers is not in the document]
Comment (gonzaloriestra) [reply packages/app/src/cli/services/generate.ts:40]: I agree. It's part of the last task here: https://github.com/Shopify/internal-cli-foundations/issues/579

File: packages/app/src/cli/services/generate.ts [hunk 4]
```diff
@@ -184,46 +167,49 @@ function formatSuccessfulRunMessage(
...
     )
   }
 
   // Validate limits for selected type.
   // If no type is selected, filter out any types that have reached their limit
-  templateSpecification.types.forEach((spec) => {
-    if (limitReached(app, spec)) {
+  templateSpecification.types.forEach((type) => {
+    if (limitReached(app, specifications, type)) {
       throw new AbortError(
```
[the remark this answers is not in the document]
Comment (gonzaloriestra) [reply packages/app/src/cli/services/generate.ts:210]: Done!

File: packages/app/src/cli/prompts/generate/extension.ts [hunk 5]
```diff
@@ -40,7 +39,7 @@ export function buildChoices(templateSpecifications: TemplateSpecification[]) {
   return templateSpecChoices.sort((c1, c2) => c1.label.localeCompare(c2.label))
```
Comment (isaacroldan) [inline packages/app/src/cli/prompts/generate/extension.ts:39]: can we force that if `spec.group` is an empty string, default to `Other`?

File: packages/app/src/cli/services/generate.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (alvaro-shopify) [reply packages/app/src/cli/services/generate.ts:101]: ok, but I guess we should handle that situation because you won't be able to deploy the app.

File: packages/app/src/cli/models/templates/theme-specifications/theme.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    {
+      url: 'https://github.com/Shopify/cli',
+      type: 'theme',
+      extensionPoints: [],
+      supportedFlavors: [
+        {
+          name: 'JavaScript',
+          value: 'vanilla-js',
+          path: 'packages/app/templates/theme-extension',
+        },
```
[the remark this answers is not in the document]
Comment (gonzaloriestra) [reply packages/app/src/cli/models/templates/theme-specifications/theme.ts:21]: I was going to ask about this. I agree JavaScript doesn't fit very well here (it's actually liquid + css + js...). I'll create a new flavor value called `liquid` if you agree.

File: packages/app/src/cli/models/templates/ui-specifications/tax_calculation.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  identifier: 'tax_calculation',
+  name: 'Tax Calculation',
+  group: 'Merchant admin',
+  supportLinks: [],
+  types: [
+    {
+      url: 'https://github.com/Shopify/cli',
+      type: 'tax_calculation',
+      extensionPoints: [],
+      supportedFlavors: uiFlavors('packages/app/templates/ui-extensions/projects/tax_calculation'),
```
[the remark this answers is not in the document]
Comment (isaacroldan) [reply packages/app/src/cli/models/templates/ui-specifications/tax_calculation.ts:17]: I guess it can't be empty, but just 1 flavor and maybe call it `tax_calculations` as well?
This extension is config only

File: packages/app/src/cli/services/generate.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (isaacroldan) [reply packages/app/src/cli/services/generate.ts:101]: Agree with @gonzaloriestra on this one,  `>=` makes sense.

File: packages/app/src/cli/prompts/generate/extension.ts [hunk 8]
```diff
@@ -69,52 +67,44 @@ const generateExtensionPrompt = async (
...
   const templateSpecification = templateSpecifications.find((spec) => spec.identifier === templateType)!
 
-  const nameAndFlavors: {name: string; flavor?: string; specification: GenericSpecification}[] = []
-  for (const spec of templateSpecification.types) {
-    // eslint-disable-next-line no-await-in-loop
-    nameAndFlavors.push(await promptNameAndFlavor(options, spec))
+  const extensionContent: GenerateExtensionContentOutput[] = []
+  /* eslint-disable no-await-in-loop */
+  for (const [index, spec] of templateSpecification.types.entries()) {
+    const name = (templateSpecification.types.length === 1 && options.name) || (await promptName(options.directory))
```
[the remark this answers is not in the document]
Comment (isaacroldan) [reply packages/app/src/cli/prompts/generate/extension.ts:78]: We are still not sure what will happen with multi-types templates. So I wouldn't invest to much in handling every case for now.

File: packages/app/src/cli/prompts/generate/extension.test.ts [hunk 9]
```diff
@@ -19,10 +18,9 @@ beforeEach(() => {
 })
 
 describe('extension prompt', async () => {
-  // ALL UI Specs, filter out theme
-  const allUISpecs = convertSpecificationsToTemplate(await loadLocalUIExtensionsSpecifications())
-  const allFunctionSpecs = testRemoteTemplateSpecifications.map(mapRemoteTemplateSpecification)
-  const allSpecs = convertSpecificationsToTemplate(await loadLocalExtensionsSpecifications())
+  const allUISpecs = localTemplateSpecifications()
+  const allFunctionSpecs = testRemoteTemplateSpecifications
+  const allSpecs = allFunctionSpecs.concat(allUISpecs).concat(themeSpecification)
```
Comment (isaacroldan) [inline packages/app/src/cli/prompts/generate/extension.test.ts:23]: doesn't `localTemplateSpecifications` already include theme?

File: packages/app/src/cli/prompts/generate/extension.test.ts [hunk 10]
```diff
@@ -34,34 +32,34 @@ describe('extension prompt', async () => {
   }
 
   test('when name is not passed', async () => {
-    const answers = {name: 'ext', extensionType: 'ui_extension'}
+    const answers = {name: 'ext', extensionType: 'checkout_ui_extension'}
```
Comment (isaacroldan) [inline packages/app/src/cli/prompts/generate/extension.test.ts:35]: why did you change this type, isn't ui_extension also a valid template id?
not saying is wrong, just curious

File: packages/app/src/cli/services/generate.ts [hunk 11]
```diff
@@ -90,31 +79,36 @@ async function buildPromptOptions(
...
   }
 }
 
 async function saveAnalyticsMetadata(promptAnswers: GenerateExtensionPromptOutput, typeFlag: string | undefined) {
   await Promise.all(
     promptAnswers.extensionContent.map((extensionContent) => {
       return metadata.addPublicMetadata(() => ({
-        cmd_scaffold_template_flavor: extensionContent.extensionFlavor,
-        cmd_scaffold_type: extensionContent.specification.identifier,
-        cmd_scaffold_type_category: extensionContent.specification.category(),
```
Comment (isaacroldan) [inline packages/app/src/cli/services/generate.ts:115]: do we want to track this somehow? categories don't apply anymore, but maybe we want to track something here?

File: packages/app/src/cli/services/generate.ts [hunk 12]
```diff
@@ -90,31 +79,36 @@ async function buildPromptOptions(
...
   }
   return groupBy(templateSpecifications, iterateeFunction)
 }
 
-function limitReached(app: AppInterface, type: GenericSpecification) {
-  if (isFunctionExtension(type)) {
+function limitReached(app: AppInterface, specifications: GenericSpecification[], templateType: TemplateType) {
+  const type = templateType.type
+  if (type === 'function') {
     return app.extensions.function.length >= blocks.functions.defaultRegistrationLimit
```
[the remark this answers is not in the document]
Comment (isaacroldan) [reply packages/app/src/cli/services/generate.ts:97]: there is a single specification called `function` that includes the global limit, since all the templates have the new type=function, it will pick that up and compare. But yeah, we can do that in a different PR!

File: packages/app/src/cli/services/generate/extension.ts [hunk 13]
```diff
@@ -78,116 +48,179 @@ function getTemplateLanguage(flavor: ExtensionFlavorValue): TemplateLanguage {
...
+      case 'theme':
+        await themeExtensionInit(options)
+        break
+      case 'function':
+        await functionExtensionInit(options)
+        break
+      default:
+        await uiExtensionInit(options)
+        break
+    }
```
[the remark this answers is not in the document]
Comment (isaacroldan) [reply packages/app/src/cli/services/generate/extension.ts:108]: ah for sure, maybe we can use the flavor as the variable to look for 🤔 
But still functions will need some custom stuff... In my extension refactor i'm adding new concepts to the specifications that might help here

**Review by isaacroldan — APPROVED** [review]
🎩 , great work @gonzaloriestra 👏

**Review by isaacroldan — APPROVED** [review]
🎩 , great work @gonzaloriestra 👏
